app/utils/service_client/middleware/body_encoder.rb

```diff
@@ -0,0 +1,90 @@
+module ServiceClient
+  module Middleware
+    class JSONEncoder
+      def encode(body)
+        body.to_json unless body.nil?
+      end
+
+      def decode(body)
+        JSON.parse(body) unless body.nil?
+      end
+
+      def mime_type
+        "application/json"
+      end
+    end
+
+    class TransitEncoder
+
+      ENCODINGS = {
+        json: "application/transit+json",
+        msgpack: "application/transit+msgpack",
+      }
+
+      def initialize(encoding)
+        @_encoding = encoding
+      end
+
+      def decode(body)
+        TransitUtils.decode(body, @_encoding) unless body.nil?
+      end
+
+      def encode(body)
+        TransitUtils.encode(body, @_encoding) unless body.nil?
+      end
+
+      def mime_type
+        ENCODINGS[@_encoding]
+      end
+    end
+
+    class BodyEncoder < MiddlewareBase
+      class ParsingError < StandardError
+      end
+
+      def initialize(encoding)
+        @_encoder = choose_encoder(encoding)
+      end
+
+      def enter(ctx)
+        req = ctx.fetch(:req)
+
+        body = req[:body]
+        headers = req.fetch(:headers)
+
+        ctx[:req][:body] = @_encoder.encode(body)
+        ctx[:req][:headers]["Accept"] = @_encoder.mime_type
+        ctx[:req][:headers]["Content-Type"] = @_encoder.mime_type unless body.nil?
+
+        ctx
+      end
+
+      def leave(ctx)
+        res = ctx.fetch(:res)
+        body = res[:body]
```

Comment:
The encoder is now chosen statically instead of using the responses content-type header. I think this is ok because we expect the API to work like that when we specifically set the Accept header ourselves. Still, I suggest adding a validation here that checks the responses content-type value and ensures it is of expected type. That will fail more meaningfully than having the decoding process raise some random parsing error.